File: web/src/api/storagevolume.ts
/**
 * 存储卷接口（F-5-02，LVM 多盘聚合）。
 *
 * 这一块最容易想错的是**条带与镜像的区别**，而两者的方向正好相反：
 *
 *   条带（stripe）是拿**可靠性**换性能——数据分散在多块盘上，吞吐更高，
 *                 但任何一块盘故障都会让整个卷不可用，而且因为数据是分散的，
 *                 剩下那些盘上的内容也无法单独恢复。
 *   镜像（mirror）是拿**容量**换可靠性——每份数据写多遍，可容忍坏盘，
 *                 代价是物理占用变成可用容量的 MirrorCount 倍。
 *
 * 界面上必须把这两句话说得同样清楚：用户看到「用了 4 块盘」时，很自然地
 * 会以为那是 4 块盘的冗余。
 */
import { del, get, post } from './client'

export type VolumeStatus = 'active' | 'sync' | 'degraded' | 'failed'

export interface VolumeView {
  id: number
  node_id: number
  name: string
  kind: string

  /** 可用容量。 */
  size_gb: number
  /** **实际占用的物理空间**——镜像会把它变成 size × mirror_count。 */
  physical_gb: number

  stripe_count: number
  mirror_count: number
  /** 为 true 表示能容忍一块盘故障。**只有镜像能。** */
  has_redundancy: boolean

  devices: string[]
  vg_name?: string
  lv_name?: string

  status: VolumeStatus
  /** 用人话解释状态，尤其是 sync 与 degraded。 */
  status_note?: string
  created_at: string
}

export interface VolumePlan {
  /** 聚合**至少**需要的设备数（stripe × mirror）。 */
  required_devices: number
  given_devices: number
  size_gb: number
  physical_gb: number
  has_redundancy: boolean
  /** 有内容时创建需要显式确认。 */
  warnings?: string[]
}

export interface VolumeRequest {
  name: string
  size_gb: number
  stripe_count?: number
  mirror_count?: number
  devices: string[]
}

export const volumeApi = {
  list: (nodeID: number) => get<{ items: VolumeView[] }>('/api/v1/storage-volumes', { node_id: nodeID }),

  /** 预检：只看会得到什么，不产生任何改动。 */
  preview: (nodeID: number, req: VolumeRequest) =>
    post<VolumePlan>(`/api/v1/storage-volumes/preview?node_id=${nodeID}`, req),

  /**
   * 创建。`acknowledge` 为 false 时若有警告，服务端**不创建也不报错**，
   * 而是把计划原样返回——那是一个需要用户做决定的岔路口。
   */
  create: (nodeID: number, req: VolumeRequest, acknowledge: boolean) =>
    post<{ plan: VolumePlan; task?: unknown }>(`/api/v1/storage-volumes?node_id=${nodeID}`, {
      ...req,
      acknowledge,
    }),

  /** 删除。**会销毁卷里的全部数据**，不可恢复。 */
  remove: (id: number) => del<{ task: unknown }>(`/api/v1/storage-volumes/${id}`),
}

export const VOLUME_STATUS_LABEL: Record<VolumeStatus, string> = {
  active: '正常',
  sync: '同步中',
  degraded: '降级',
  failed: '故障',
}

export const VOLUME_STATUS_TONE: Record<VolumeStatus, 'success' | 'warning' | 'danger'> = {
  active: 'success',
  // sync 用 warning 而不是 success：它确实可用，但写会明显变慢几小时。
  sync: 'warning',
  // degraded 用 danger：它**看起来一切正常**却已经没有任何冗余了。
  degraded: 'danger',
  failed: 'danger',
}

File: internal/agent/storage_volume.go
package agent

// OpStorageVolumeApply 创建或删除一个 LVM 存储卷（F-5-02）。
//
// 一次操作覆盖创建与删除，由 `action` 区分——两者的参数几乎一样（都只要
// 名称与设备），而拆成两个 OpKind 只会让「哪些参数属于哪一边」这件事需要
// 在两处各维护一遍。
const OpStorageVolumeApply OpKind = "storage.volume.apply"

// VolumeDataKey 是结果中承载补充信息的键。
const VolumeDataKey = "storage_volume"

// VolumeInfo 是节点回传的存储卷信息。
type VolumeInfo struct {
	// Status 是节点侧判断的卷状态（active / sync / degraded / failed）。
	//
	// **由节点给出而不是控制面一律写 active**：镜像卷建好之后还要同步
	// 几小时，期间读正常、写明显变慢。把这段时间标成 active，用户会认为
	// 卷是好的、慢是别的原因——于是一路排查到网络。
	Status string
	// SizeGB 与 PhysicalGB 是节点侧算出的容量。
	//
	// 与控制面的换算**分开返回**：两者不一致说明聚合方式与预期不符
	// （例如镜像份数没有生效），而那种问题如果只信控制面自己算的数字，
	// 就永远发现不了——它会一直显示「占用翻倍」而实际并没有冗余。
	SizeGB     int
	PhysicalGB int
	Message    string
	// Warnings 是节点发现的、值得用户先看一眼的问题。
	Warnings []string
}
